data/code/lib.c

```diff
@@ -589,3 +593,715 @@ void check_lap_ex()
+#endif
+    int* pointer_pointer = (int*)(*race_manager_pointer + 0x38);
+    char* mirror_flag = (char*)(*pointer_pointer + 0x2c);
+    return *mirror_flag;
+}
+
+// Returns simplified number for easy determination of stick position.
+signed int get_stick_dir_id(char* const ctrl, char* const race_manager, int kart_num)
+{
+    char stick = get_stick_ctrl(ctrl, kart_num);
+    signed int ret = 0;
+    if ((stick & 0x1) != 0)  // Right.
+    {
+        ret = -1;
+    }
+    else if ((stick & 0x2) != 0)  // Left.
+    {
+        ret = 1;
+    }
+    if (is_mirror(race_manager) == false)  // Flip if NOT mirror.
+    {
+        ret *= -1;
+    }
+    return ret;
+}
+
+// Main function for shifting sideways during bounce.
+void handle_x_adjustment(char* const this, char* const ctrl, char* const race_manager, int kart_num)
+{
+    float* const last_momentum = &s_last_momenta[kart_num];
+
+    float z_direction_vector[] = {0.0, 0.0, 0.0};
+    ObjUtility__getKartZdir(kart_num, z_direction_vector);
+
+    signed int stick_dir_id = get_stick_dir_id(ctrl, race_manager, kart_num);
+
+    float speed = 0;
+
+    if (stick_dir_id != 0)  // If holding left or right.
+    {
+        speed = add_speed(*last_momentum, stick_dir_id);
+    }
+    else
+    {
+        speed = deaccelerate_speed(*last_momentum);
+    }
+    *last_momentum = speed;  // Stores to 0x80005240 + kart_num.
+    speed *= 10.0;
+
+    z_direction_vector[0] *= speed;
+    z_direction_vector[2] *= speed;
+    add_absolute_position_vector(
+        this, z_direction_vector[0], z_direction_vector[1], z_direction_vector[2]);
+}
+
+// Resets last recorded XZ momentum before bounce liftoff;
+void reset_last_momentum(int kart_num)
+{
+    float* const last_momentum = &s_last_momenta[kart_num];
+    *last_momentum = 0;
+}
+
+void call_do_spd_ctrl_normal(char* const strat)
+{
+    KartStrat__DoSpeedCrl(strat);  // Original instruction;
+}
+
+void call_do_spd_ctrl_mod(char* const this,
+                          char* const ctrl,
+                          char* const race_manager,
+                          int kart_num)
+{
+    handle_boosts(this);
+    handle_x_adjustment(this, ctrl, race_manager, kart_num);
+    handle_y_adjustment(this, ctrl, kart_num);
+    clamp_movement_vector_descent(this);
+}
+
+void call_do_spd_ctrl(char* const this,
+                      char* const strat,
+                      char* const ctrl,
+                      char* const race_manager,
+                      int kart_num,
+                      int kart_bounce_flag)
+{
+    if (kart_bounce_flag == false)
+    {
+        call_do_spd_ctrl_normal(strat);
+    }
+    else
+    {
+        call_do_spd_ctrl_mod(this, ctrl, race_manager, kart_num);
+
+        // ASM section done here instead of its own function (has differing results?).
+        // Forcefully set registers to result of original instruction for peace of mind.
+
+        asm("lis %r4, 0x8143");
+        asm("ori %r4, %r4, 0x3a50");
```

Comment:
Assuming this is an address (`0x81433A50`), it appears to be pointing to head memory. I may be wrong, but hardcoding those addresses is almost always wrong: it won't work if there is anything in the game that would allocate _more_ or _less_ memory.

As a matter of fact, by virtue of shifting the OS arena [which the Extender does depending of the size of the compiled code that needs to be injected], all addresses in the heap will be shifted.

Happy to hear your thoughts on this.

Reply:
To be perfectly honest, I did that without considering what it does. It just happened to be in that register when I was looking at how the registers end up at the end of DoSpeedCrl, and threw it in without any further thought. It's almost certainly unnecessary - honestly, all of the asm writes in that section probably are.

Reply:
I decided to remove this asm section.